File: test/_harnesses/go/multiple_broken/multiple_broken.go
package main

import (
	"encoding/binary"
	"flag"
	"io"
	"os"
	"path/filepath"

	"messages/broken_messages"
)

var ok bool = true

func softAssert(condition bool, label string) {
	if !condition {
		os.Stderr.WriteString("FAILED! Go: " + label + "\n")
		ok = false
	}
}

func main() {
	var trunc broken_messages.TruncatedMessage
	trunc.X = 1.0
	trunc.Y = 2.0

	var full broken_messages.FullMessage
	full.X = 1.0
	full.Y = 2.0
	full.Z = 3.0

	generatePathPtr := flag.String("generate", "", "")
	readPathPtr := flag.String("read", "", "")
	flag.Parse()

	if len(*generatePathPtr) > 0 {
		os.MkdirAll(filepath.Dir(*generatePathPtr), os.ModePerm)
		dat, err := os.Create(*generatePathPtr)
		if err != nil {
			panic(err)
		}
		defer dat.Close()

		full.WriteBytes(dat, true)
		full.WriteBytes(dat, true)
		full.WriteBytes(dat, true)

		// write a truncated message tagged as a full one
		binary.Write(dat, binary.LittleEndian, broken_messages.FullMessageType)
		trunc.WriteBytes(dat, false)

		full.WriteBytes(dat, true)
		full.WriteBytes(dat, true)
		full.WriteBytes(dat, true)

		size := 6 * full.GetSizeInBytes()
		size += 6 // markers, one byte each
		size += trunc.GetSizeInBytes()
		size += 1 // trunc marker

		seek, _ := dat.Seek(0, io.SeekCurrent)
		softAssert(size == (int)(seek), "written bytes check")
	} else if len(*readPathPtr) > 0 {
		dat, err := os.Open(*readPathPtr)
		if err != nil {
			panic(err)
		}
		defer dat.Close()

		_, err = broken_messages.ProcessRawBytes(dat)
		softAssert(err != nil, "read broken stream")
		softAssert(err.Error() == "Unknown message type: 63", "broken stream error message")
	}

	if !ok {
		os.Stderr.WriteString("Failed assertions.\n")
		os.Exit(1)
	}
}

File: test/_harnesses/assemblyscript/assembly/multiple_broken.ts
import * as harness from "./_harness";
export * from "./_harness";

import * as BrokenMessages from '../../../../out/generated/assemblyscript/BrokenMessages';

const trunc = new BrokenMessages.TruncatedMessage();
trunc.x = 1.0;
trunc.y = 2.0;

const full = new BrokenMessages.FullMessage();
full.x = 1.0;
full.y = 2.0;
full.z = 3.0;

export function generate(): usize {
    let size: usize = 6 * full.getSizeInBytes();
    size += 6; // markers, one byte each
    size += trunc.getSizeInBytes();
    size += 1; // trunc marker

    harness.allocate(size);
    const dv = harness.getDataView();
    const da = new BrokenMessages.DataAccess(dv);

    let success: bool = true; // good vibes
    success = full.writeBytesDA(da, true);
    success = full.writeBytesDA(da, true);
    success = full.writeBytesDA(da, true);

    // write a truncated message tagged as a full one
    da.setByte(BrokenMessages.MessageType.FullMessageType as u8);
    trunc.writeBytesDA(da, false);

    success = full.writeBytesDA(da, true);
    success = full.writeBytesDA(da, true);
    success = full.writeBytesDA(da, true);

    return harness.getDataPtr();
}

export function read(): void {
    const dv = harness.getDataView();

    const msgList = BrokenMessages.ProcessRawBytes(dv, -1);

    harness.softAssert(msgList.length == 4, "read broken stream")
}
